render_test.go

```diff
@@ -10,45 +10,77 @@ import (
 
 	"github.com/go-text/render"
 	"github.com/go-text/typesetting/font"
+	"github.com/go-text/typesetting/shaping"
+
+	"golang.org/x/image/math/fixed"
 )
 
 func Test_Render(t *testing.T) {
-	img := image.NewNRGBA(image.Rect(0, 0, 420, 180))
+	img := image.NewNRGBA(image.Rect(0, 0, 425, 250))
 	draw.Draw(img, img.Bounds(), &image.Uniform{color.White}, image.Point{}, draw.Src)
 
 	data, _ := os.Open("testdata/NotoSans-Regular.ttf")
-	f, _ := font.ParseTTF(data)
+	f1, _ := font.ParseTTF(data)
 
 	r := &render.Renderer{
 		FontSize: 48,
 		Color:    color.Black,
 	}
 	str := "Hello! ± ज्या"
-	r.DrawString(str, img, f)
-	r.DrawStringAt(str, img, 0, 100, f)
+	r.DrawString(str, img, f1)
+	r.DrawStringAt(str, img, 0, 100, f1)
 
 	r.PixScale = 2
 	r.Color = color.Gray{Y: 0xcc}
-	r.DrawStringAt("baseline", img, 0, 180, f)
+	r.DrawStringAt("baseline", img, 0, 180, f1)
 
 	data, _ = os.Open("testdata/NotoSans-Bold.ttf")
-	f, _ = font.ParseTTF(data)
+	f2, _ := font.ParseTTF(data)
 	r.FontSize = 36
 	r.Color = color.NRGBA{R: 0xcc, G: 0, B: 0x33, A: 0x99}
-	x := r.DrawStringAt("Red", img, 60, 140, f)
-	r.DrawStringAt("Bold", img, x, 140, f)
+	x := r.DrawStringAt("Red", img, 60, 140, f2)
+	r.DrawStringAt("Bold", img, x, 140, f2)
 
 	// from https://github.com/adobe-fonts/emojione-color, MIT license
 	data, _ = os.Open("testdata/EmojiOneColor.otf")
-	f, _ = font.ParseTTF(data)
+	f3, _ := font.ParseTTF(data)
 	r.FontSize = 36
-	r.DrawStringAt("😀🎁", img, 270, 80, f)
+	r.DrawStringAt("🚀🖥️", img, 270, 80, f3)
 
 	data, _ = os.Open("testdata/Greybeard-22px.ttf")
-	f, _ = font.ParseTTF(data)
+	f4, _ := font.ParseTTF(data)
 	r.FontSize = 22
 	r.Color = color.NRGBA{R: 0xcc, G: 0x66, B: 0x33, A: 0xcc}
-	r.DrawStringAt("\uE0A2░", img, 362, 149, f)
+	r.DrawStringAt("\uE0A2░", img, 366, 164, f4)
+
+	str = "Hello ज्या 😀! 🎁 fin."
+	rs := []rune(str)
+	sh := &shaping.HarfbuzzShaper{}
+	in := shaping.Input{
+		Text:     rs,
+		RunStart: 0,
+		RunEnd:   len(rs),
+		Size:     fixed.I(int(r.FontSize)),
+	}
+	faces := []font.Face{f1, f2, f3}
+	fontRuns := shaping.SplitByFontGlyphs(in, faces)
+	seg := shaping.Segmenter{}
+
+	x = 0
+	for _, funtRun := range fontRuns {
+		runs := seg.Split(funtRun, singleFontMap{funtRun.Face})
```

Comment:
This looks strange to me. 
If I'm not mistaken, you can directly use `seg.Split` (or even better r.segmenter.Split, reusing the cached segmenter), and remove the outer loop.
This will require to use a slighty more advanced `FontMap` concrete implementation, like the one at https://github.com/go-text/typesetting/blob/main/shaping/input.go#L78

By the way, perhaps we should expose `shaping.fixedFontmap`.

Reply:
To be explicit, I would write something like  

```go
seg := shaping.Segmenter{}
runs := seg.Split(in, fixedFontmap(faces))

line := make(shaping.Line, len(runs))
for i, run := range runs {
    line[i] = sh.Shape(run)
}

r.Color = color.NRGBA{R: 0x33, G: 0x99, B: 0x33, A: 0xcc}
x = 0
for _, run := range line {
     x = r.DrawShapedRunAt(run, img, x, 232)
}
```

where `fixedFontmap` is defined by 

```go
type fixedFontmap []font.Face

// ResolveFace panics if the slice is empty
func (ff fixedFontmap) ResolveFace(r rune) font.Face {
	for _, f := range ff {
		if _, has := f.NominalGlyph(r); has {
			return f
		}
	}
	return ff[0]
}
```

Reply:
> ```go
> fixedFontmap(faces)
> ```

Perfect thanks, I have an updated commit with that content - much simpler.

> (or even better r.segmenter.Split, reusing the cached segmenter)

It is currently private, so no access. I don't think it makes sense to add public APIs for the sake of re-use here does it? if there is one external and one internal that's not a big deal.

Reply:
> It is currently private, so no access. I don't think it makes sense to add public APIs for the sake of re-use here does it? if there is one external and one internal that's not a big deal.

Oh yes, you're right ! I was thinking of an internal implementation, but for this demo this is just fine !